trie/utils/verkle.go

```diff
@@ -0,0 +1,302 @@
+	// - poly[3]'s byte representation is the same as the *top* 16 bytes (trieIndexBytes[16:]) of
+	//   32-byte aligned big-endian representation (BE({00,...,AA,BB,CC})).
+	// - poly[4]'s byte representation is the same as the *low* 16 bytes (trieIndexBytes[:16]) of
+	//   the 32-byte aligned big-endian representation (BE({00,00,...}).
+	trieIndexBytes := treeIndex.Bytes32()
+	verkle.FromBytes(&poly[3], trieIndexBytes[16:])
+	verkle.FromBytes(&poly[4], trieIndexBytes[:16])
+
+	cfg := verkle.GetConfig()
+	ret := cfg.CommitToPoly(poly[:], 0)
+
+	// add a constant point corresponding to poly[0]=[2+256*64].
+	ret.Add(ret, getTreePolyIndex0Point)
+
+	return pointToHash(ret, subIndex)
+}
+
+// GetTreeKeyAccountLeaf returns the key to the first leaf of
+// an account.
+func GetTreeKeyAccountLeaf(address []byte, leaf byte) []byte {
+	return GetTreeKey(address, zero, leaf)
+}
+
+func GetTreeKeyVersion(address []byte) []byte {
+	return GetTreeKey(address, zero, VersionLeafKey)
+}
+
+// Get does the,
+// but uses the cached point corresponding to the address part of the polynomial.
+// GetTreeKeyVersionWithEvaluatedAddress does the same thing as GetTreeKeyVersion,
+// but reuses a cached evaluation of the address part of the polynomial.
+func GetTreeKeyVersionWithEvaluatedAddress(addrp *verkle.Point) []byte {
+	return getTreeKeyWithEvaluatedAddess(addrp, zero, VersionLeafKey)
+}
+
+func GetTreeKeyBalance(address []byte) []byte {
+	return GetTreeKey(address, zero, BalanceLeafKey)
+}
+
+func GetTreeKeyNonce(address []byte) []byte {
+	return GetTreeKey(address, zero, NonceLeafKey)
+}
+
+func GetTreeKeyCodeKeccak(address []byte) []byte {
+	return GetTreeKey(address, zero, CodeKeccakLeafKey)
+}
+
+func GetTreeKeyCodeSize(address []byte) []byte {
+	return GetTreeKey(address, zero, CodeSizeLeafKey)
+}
+
+func GetTreeKeyCodeChunk(address []byte, chunk *uint256.Int) []byte {
+	chunkOffset := new(uint256.Int).Add(CodeOffset, chunk)
+	treeIndex := new(uint256.Int).Div(chunkOffset, VerkleNodeWidth)
+	subIndexMod := new(uint256.Int).Mod(chunkOffset, VerkleNodeWidth)
+	var subIndex byte
+	if len(subIndexMod) != 0 {
+		subIndex = byte(subIndexMod[0])
+	}
+	return GetTreeKey(address, treeIndex, subIndex)
+}
+
+// GetTreeKeyCodeChunkWithEvaluatedAddress does the same thing as GetTreeKeyCodeChunk,
+// but uses the cached point corresponding to the address part of the polynomial.
+func GetTreeKeyCodeChunkWithEvaluatedAddress(addressPoint *verkle.Point, chunk *uint256.Int) []byte {
+	chunkOffset := new(uint256.Int).Add(CodeOffset, chunk)
+	treeIndex := new(uint256.Int).Div(chunkOffset, VerkleNodeWidth)
+	subIndexMod := new(uint256.Int).Mod(chunkOffset, VerkleNodeWidth)
+	var subIndex byte
+	if len(subIndexMod) != 0 {
+		subIndex = byte(subIndexMod[0])
+	}
+	return getTreeKeyWithEvaluatedAddess(addressPoint, treeIndex, subIndex)
+}
+
+func GetTreeKeyStorageSlot(address []byte, storageKey *uint256.Int) []byte {
+	pos := storageKey.Clone()
```

Comment:
Perhaps `pos := new(uint256.Int)`? We either initialize the pos via `pos.Add(HeaderStorageOffset, storageKey)` or `pos.Add(MainStorageOffset, storageKey)`. Make no sense to do clone first.